system/src/ledger/ledger_manager.cpp

```diff
@@ -0,0 +1,1599 @@
+        setPendingState(curCtx_, PendingState::GET_INFO | PendingState::SYNC_TO_CLOUD);
+    }
+    curCtx_->taskRunning = false;
+    curCtx_ = nullptr;
+    state_ = State::READY;
+    return 0;
+}
+
+int LedgerManager::receiveGetDataResponse(CoapMessagePtr& msg, int result) {
+    assert(state_ == State::SYNC_FROM_CLOUD && curCtx_ && curCtx_->syncDir == LEDGER_SYNC_DIRECTION_CLOUD_TO_DEVICE &&
+            curCtx_->taskRunning && !stream_ && !msg_);
+    if (result != 0) {
+        LOG(ERROR, "Failed to sync ledger: %s; result: %d", curCtx_->name, result);
+        if (!isLedgerAccessError(result)) {
+            return SYSTEM_ERROR_LEDGER_REQUEST_FAILED;
+        }
+        // Ledger may no longer be accessible, re-request its info
+        LOG(WARN, "Re-requesting ledger info: %s", curCtx_->name);
+        setPendingState(curCtx_, PendingState::GET_INFO | PendingState::SYNC_FROM_CLOUD);
+        curCtx_->taskRunning = false;
+        curCtx_ = nullptr;
+        state_ = State::READY;
+        return 0;
+    }
+    LOG(TRACE, "Received ledger data: %s", curCtx_->name);
+    // Open the ledger for writing
+    RefCountPtr<Ledger> ledger;
+    CHECK(getLedger(ledger, curCtx_->name));
+    std::unique_ptr<LedgerWriter> writer(new(std::nothrow) LedgerWriter());
+    CHECK(ledger->initWriter(*writer, LedgerWriteSource::SYSTEM));
+    // Ledger data may span multiple CoAP messages. Nanopb streams are synchronous so the response
+    // is decoded manually
+    pb_istream_t pbStream = {};
+    CHECK(getStreamForSubmessage(msg.get(), &pbStream, PB_CLOUD(Response_ledger_get_data_tag)));
+    uint64_t lastUpdated = 0;
+    for (;;) {
+        uint32_t fieldTag = 0;
+        auto fieldType = pb_wire_type_t();
+        bool eof = false;
+        if (!pb_decode_tag(&pbStream, &fieldType, &fieldTag, &eof)) {
+            if (!eof) {
+                return SYSTEM_ERROR_BAD_DATA;
+            }
+            break;
+        }
+        if (fieldTag == PB_LEDGER(GetDataResponse_last_updated_tag)) {
+            if (fieldType != PB_WT_64BIT || !pb_decode_fixed64(&pbStream, &lastUpdated)) {
+                return SYSTEM_ERROR_BAD_DATA;
+            }
+        } else if (fieldTag == PB_LEDGER(GetDataResponse_data_tag)) {
+            if (!pb_skip_field(&pbStream, PB_WT_VARINT)) { // Skip the field length
+                return SYSTEM_ERROR_BAD_DATA;
+            }
+            // "data" is always the last field in the message. XXX: It's assumed that the message
+            // fields are encoded in order of their field numbers, which is not guaranteed by the
+            // Protobuf spec in general
+            break;
+        }
+    }
+    writer->updateInfo(LedgerInfo().lastUpdated(lastUpdated));
+    stream_.reset(writer.release());
+    msg_ = std::move(msg);
+    // Read the first chunk of the ledger data
+    CHECK(receiveLedgerData());
+    return 0;
+}
+
+int LedgerManager::receiveSubscribeResponse(CoapMessagePtr& msg, int result) {
```

Comment:
What would happen if a response went missing? For example, you send a subscribe request and the server sends the subscribe response but it gets lost along the way. Is the timeout mechanism at the CoAP layer? I noticed the calls to `coap_begin_request` have timeout set to 0. I'm not sure if it means "use default timeout" or "no timeout".